File: Leetcode/Python3/328. Odd Even Linked List/solution.py
# https://leetcode.com/problems/odd-even-linked-list/

# Ref: https://leetcode.com/problems/odd-even-linked-list/discuss/133345/With-detailed-explanation-or-Python

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head == None:    # Avoid blank list
            return head
        odd = head
        even = head.next
        even_head = even    # Save even index head node
        while even and even.next:
            odd.next = odd.next.next
            even.next = even.next.next
            odd = odd.next
            even = even.next
        odd.next = even_head    # Link odd_index linked list to even_index linked list head
        return head

File: Leetcode/Kotlin/328. Odd Even Linked List/solution.kt
// https://leetcode.com/problems/odd-even-linked-list/submissions/

/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */

//  TODO: fix null check
class Solution {
    fun oddEvenList(head: ListNode?): ListNode? {
        if (head == null){
            return null;
        }

        val evenHead = head.next
        var even = evenHead
        var odd = head

        while (even?.next != null){
            odd?.next = odd?.next?.next
            even?.next = even?.next?.next

            even = even?.next
            odd = odd?.next
        }

        odd?.next = evenHead

        return head
    }
}